### PR TITLE
Add a cli-based ChannelContext

### PR DESCRIPTION
**Pull request by jakubtobiasz**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13                   |
| Bug fix?        | no                                                       |
| New feature?    | yes                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | no  |
| Related tickets | #9987                      |
| License         | MIT                                                          |

Hello,
This PR is simple, but I spend some time on trying different approaches and this is a result. This PR only provides a way to allow defining a channel via console command. Once approved, I will add a cookbook how to create a channel aware command.

I tried some approach, then I read comments from #9987. 

- Interfaces/traits → I didn't use them intentionally, the input value can be achieved only through `execute()` method so it might be hard to push the value to be available in another method
- Subscribers/Event Listener → We can listen on the command event, then get a value and push it into some field (e.g., using the above approach with interfaces/traits), but first things is we do a little magic to the solution, another thing we hard code 'channel' option or any other name we pick

I decided to create a some kind of bag we can push channel code we want to use inside a command, then it is available everywhere where we use channel context. I'm open to any suggestions 🤷🏼‍♂️.

### CODE REVIEW

**Comment by vvasiloi** [comment]
Your proposal is good, but it solves only part of the problem and leaves the rest to the users.
The point of the interface, trait and listener was to spare the user all the boilerplate of adding the channel to the input, validating it, setting it and maybe even unsetting it.

**Comment by lchrusciel** [comment]
What would you say about the introduction of `SYLIUS_CHANNEL_ENV` and the new channel context `EnvAwareChannelContext`, which will receive the value from the CLI into the constructor, may have quite high priority, by default null, so it will fallback to the current logic. 

This solution will drop boilerplate code, and reuse Symfony features and we won't be forced to do any adjustments in the currently working code. What is more, it will provide a new feature - if you would like to force Sylius to use given channel for given endpoint, you can define this env withing nginx / apache

**Comment by vvasiloi** [comment]
@lchrusciel genius.

**Comment by jakubtobiasz** [comment]
I updated the PR according to suggestions from @lchrusciel. I'm open for more suggestions.

File: src/Sylius/Bundle/ChannelBundle/Resources/config/app/config.yml [hunk 1]
```diff
@@ -1,6 +1,9 @@
 # This file is part of the Sylius package.
 # (c) Paweł Jędrzejewski
 
+parameters:
+    sylius_channel_env: "%env(default::SYLIUS_CHANNEL)%"
```
Comment (vvasiloi) [inline src/Sylius/Bundle/ChannelBundle/Resources/config/app/config.yml]: This parameter is redundant
```suggestion
    env(SYLIUS_CHANNEL): ~
```

File: src/Sylius/Bundle/ChannelBundle/Resources/config/services.xml [hunk 2]
```diff
@@ -40,6 +40,12 @@
         <service id="sylius.context.channel.composite" class="Sylius\Component\Channel\Context\CompositeChannelContext" public="false" />
         <service id="Sylius\Component\Channel\Context\ChannelContextInterface" alias="sylius.context.channel" />
 
+        <service id="sylius.context.channel.env_based" class="Sylius\Component\Channel\Context\EnvBased\ChannelContext" public="false">
+            <argument type="service" id="sylius.repository.channel" />
+            <argument>%sylius_channel_env%</argument>
```
Comment (vvasiloi) [inline src/Sylius/Bundle/ChannelBundle/Resources/config/services.xml]:
```suggestion
            <argument>%env(SYLIUS_CHANNEL)%</argument>
```

File: src/Sylius/Component/Channel/Context/EnvBased/ChannelContext.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+{
+    public function __construct(
+        private ChannelRepositoryInterface $channelRepository,
+        private ?string $channelCodeFromEnv,
+    ) {
+    }
+
+    public function getChannel(): ChannelInterface
+    {
+        if ('cli' !== PHP_SAPI || null === $this->channelCodeFromEnv) {
```
Comment (NoResponseMate) [inline src/Sylius/Component/Channel/Context/EnvBased/ChannelContext.php:31]: Since it'll work virtually always, is there any point in limiting it only to cli?

File: src/Sylius/Component/Channel/Context/EnvBased/ChannelContext.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+declare(strict_types=1);
+
+namespace Sylius\Component\Channel\Context\EnvBased;
+
+use Sylius\Component\Channel\Context\ChannelContextInterface;
+use Sylius\Component\Channel\Context\ChannelNotFoundException;
+use Sylius\Component\Channel\Model\ChannelInterface;
+use Sylius\Component\Channel\Repository\ChannelRepositoryInterface;
+
+final class ChannelContext implements ChannelContextInterface
```
Comment (vvasiloi) [inline src/Sylius/Component/Channel/Context/EnvBased/ChannelContext.php:21]: Is it really necessary to have `EnvBased` in namespace instead of class name, like the `SingleChannelContext`?

File: src/Sylius/Component/Channel/Context/EnvBased/ChannelContext.php (hunk 3 above)
Comment (vvasiloi) [reply src/Sylius/Component/Channel/Context/EnvBased/ChannelContext.php:31]: Yes, otherwise you risk it being used in http context. The env var might be defined globally either by accident or on purpose.

Comment (jakubtobiasz) [reply src/Sylius/Component/Channel/Context/EnvBased/ChannelContext.php:31]: @vvasiloi is right, it was made intentionally to prevent mistakes, also I don't see any use-case outside the CLI context.

File: src/Sylius/Bundle/ChannelBundle/Resources/config/services.xml (hunk 2 above)
Comment (jakubtobiasz) [reply src/Sylius/Bundle/ChannelBundle/Resources/config/services.xml]: I used this way instead of the parameter.

File: src/Sylius/Component/Channel/Context/EnvBased/ChannelContext.php (hunk 4 above)
Comment (jakubtobiasz) [reply src/Sylius/Component/Channel/Context/EnvBased/ChannelContext.php:21]: Any ideas? I thought if `EnvBased` is an accurate name 🤔. We could pass anything in the constructor, and the channel is set 🤔.

Comment (vvasiloi) [reply src/Sylius/Component/Channel/Context/EnvBased/ChannelContext.php:21]: `Sylius\Component\Channel\Context\EnvBasedChannelContext` should be enough 🤷‍♂️

Comment (vvasiloi) [reply src/Sylius/Component/Channel/Context/EnvBased/ChannelContext.php:21]: It's better for DX IMO.

**Comment by diimpp** [comment]
> What would you say about the introduction of `SYLIUS_CHANNEL_ENV` and the new channel context `EnvAwareChannelContext`, which will receive the value from the CLI into the constructor, may have quite high priority, by default null, so it will fallback to the current logic.
> 
> This solution will drop boilerplate code, and reuse Symfony features and we won't be forced to do any adjustments in the currently working code. What is more, it will provide a new feature - if you would like to force Sylius to use given channel for given endpoint, you can define this env withing nginx / apache

@lchrusciel How will it work if I have multiple channels and I want different services to work with different channels?

I still fail to see any value to this feature aside from newcommers misuse of contexts in cli.
This will promote even more state-full implementations than necessary.

**Comment by vvasiloi** [comment]
@diimpp it's meant to run background jobs that use services that depend on channel context.
For example, in one project I have a command that runs as a cronjob, gets all unpaid orders older than 30 min and sends an email to customer reminding them to pay. The time limit is stored in property on channel and this is how the command starts:
```php
$channel = $this->channelContext->getChannel();
$orders = $this->orderRepository->findForOnlinePaymentNotificationSince(
    new DateTime(sprintf('- %s minutes', $channel->getPaymentNotifyAfterMinutes()))
);
```
In a multi-channel environment this command won't work, but with this new feature it will work like this:
```
*/30 * * * * SYLIUS_CHANNEL=channel1 /app/bin/console app:orders:send-payment-notification
*/30 * * * * SYLIUS_CHANNEL=channel2 /app/bin/console app:orders:send-payment-notification
```

**Comment by diimpp** [comment]
@vvasiloi 

tldr:
- Avoid global state at all cost. :)
- In http request, app has global state and channel/locale/currency contexts are brilliant design to capture it and offer use of it. There is no such thing in cli, any use of sylius contexts will be hacks, this can't become canon at sylius/sylius.
- Problem of accessing channel/locale/currency in cli is real, but there are different requirements -- one/many *different* channels to some services, not one channel to all services.
- This is not about cli commands only, but workers and services in cli context.

Imagine I've
- Channel A
- Channel B
- Channel C

then I will have
- Service1 is channel interchangeable, which works per channel A, B, C, ...
- Service2, which works only with channel A
- Service3, which works only with channel B, C
- Service4 is not channel aware.

Official solution with BC promise and years in life cycle should accommodate all those cases.

---

>```
> $channel = $this->channelContext->getChannel();
> $orders = $this->orderRepository->findForOnlinePaymentNotificationSince(
>     new DateTime(sprintf('- %s minutes', $channel->getPaymentNotifyAfterMinutes()))
> );
> ```

This is a single channel app example. This code will break if you will create second channel and some of orders will belong to first channel and some to another.

```php
$channels = $this->channelRepository->findBy(['enabled' => true, 'notifyAbandonedCarts' => true]);
foreach ($channels as $channel) {
    $orders = $this->orderRepository->findForOnlinePaymentNotificationSince(
        $channel,
        new DateTime(sprintf('- %s minutes', $channel->getPaymentNotifyAfterMinutes()))
    );
}
```


> 
> In a multi-channel environment this command won't work, but with this new feature it will work like this:
> 
> ```
> */30 * * * * SYLIUS_CHANNEL=channel1 /app/bin/console app:orders:send-payment-notification
> */30 * * * * SYLIUS_CHANNEL=channel2 /app/bin/console app:orders:send-payment-notification
> ``` 

Order based commands should work with related channel/locale/currency and not look for context channel at all. The only feasible solution here is to use cli context channel as filter for orders and this is misuse of context already and perfectly solvable on the cli command level by checking abandoned cart settings per channel.

---

This is my case exactly, I've workers, scheduled jobs and some of them work by different channels, not a swap-able channel, but one job is only for certain channel and never for other.

What if some of them work by several at once **?**

Context is just not a applicable concept for cli, because there is **no real "context", you manually provide channel or multiple channels to certain services, there is no way to detect channel from app's global state.**
This is just a convenient way to provide default single channel in familiar sylius terms, but this so wrong on so many levels.

Imagine next case: app has single channel and developer implements command/command handler and injects channel context to command handler instead of providing channel code via command. And it will work correctly, because channel is the same in http and in the cli, but then they introduce new channel and there will be errors between channels now.

That's what I mean by it will promote statefullness and bad design choices for user-developers.

---

> */30 * * * * SYLIUS_CHANNEL=channel2 /app/bin/console app:some-action

You still can have this, design similar to contexts system, but with emphasis on providing one/multiple channels to certain services, not a single channel to all services.

**Comment by GSadee** [comment]
I thought it would be a simple merge at this stage, after discussions, but I have some doubts as to whether we want to introduce it or not 😅.

I mostly agree with the above comment from @diimpp that contexts should not be applicable to CLI commands. Primarily because we are not really in a specific context when executing a command and this may in the long run introduce more unpredictable events than it benefits.

If I see it correctly, we do not use contexts in the CLI commands, either directly or through dependent services. It seems to me that if someone wants to use context in a command, they should instead explicitly provide the relevant channels, or possibly implement a similar context themselves in the end application.

By introducing this code into Sylius, we would solve the potential problem (from the initial issue) of command customisation/creating new ones, but we would also encourage more people to use contexts in this way, in my opinion incorrectly. 

The more I think about this problem, I get the feeling that we shouldn't change anything too much in the current Sylius code according to this topic 🤔.

**Comment by vvasiloi** [comment]
@GSadee this features is not for explicitely using channel context in the cli command, it's for using other services from Sylius and plugins that depend on the channel context.

**Comment by diimpp** [comment]
@GSadee  Yep, official solution would be nice, but it probably will take form of non-sylius specific domain oriented cli data provider.

Any DDD adepts here? Should it be called some kind of mixin maybe? :)

---

I was able to find few articles to support my case since then.

https://stackoverflow.com/questions/986865/can-you-explain-the-context-design-pattern
https://www.javaguides.net/2018/08/context-object-design-pattern-in-java.html

It's also my understanding, that `Context object` is considered somewhat an anti-pattern due to tendency to make a global key-value storage out of it.

**Comment by diimpp** [comment]
> @GSadee this features is not for explicitely using channel context in the cli command, it's for using other services from Sylius and plugins that depend on the channel context.

Then we should fix those services to be stateless and accept channel/locale/user/... as method argument.

**Comment by vvasiloi** [comment]
> > @GSadee this features is not for explicitely using channel context in the cli command, it's for using other services from Sylius and plugins that depend on the channel context.
> 
> Than we should fix those services to be stateless and accept channel/local/user/... as method argument.

We should, but that probably requires a lot of effort. We need to deprecate all the contexts and refactor all the services that use them.

File: src/Sylius/Component/Channel/Context/EnvBased/ChannelContext.php (hunk 3 above)
Comment (lchrusciel) [reply src/Sylius/Component/Channel/Context/EnvBased/ChannelContext.php:31]: I see the potential for it - if you would like to route your websites to given channels based on your server configuration rather than hostname resolution in the store.

**Comment by GSadee** [comment]
> @GSadee this features is not for explicitely using channel context in the cli command, it's for using other services from Sylius and plugins that depend on the channel context.

Yup, I understand that, but still, If I see it correctly, we do not use services that depend on contexts in the CLI commands in Sylius. And if someone uses them this way, then this kind of problem should be resolved there, in the end application or plugin, by refactoring CLI command and/or services or just by adding a similar context as has been introduced in this PR. IMO it should not be fixed here, in Sylius, where this additional context will not even be used.

**Comment by lchrusciel** [comment]
After a long internal discussion, we've decided that:
* This feature won't be merged
* We will provide a cookbook with this code, so every plugin that has issues with it can implement on its own
* We will provide recommendations, on how to solve channel-based issues (based on order, customer and something from the community like CMS)
* Perhaps, we will start storing information about the first contact channel on the Customer
* We will fix our own contexts

We don't want to merge it, as no Sylius CLI command is suffering from this problem. What is more, we do not consider such usage as a good solution to the problem, so we don't want to encourage others to solve it this way. Even if we merged it, this class would be instantly deprecated

**Comment by jakubtobiasz** [comment]
According to Łukasz's comment, I'm closing this PR.
